checker.c

```diff
@@ -1,21 +1,80 @@
 #include <stdio.h>
-#include <assert.h>
-
-int batteryIsOk(float temperature, float soc, float chargeRate) {
-  if(temperature < 0 || temperature > 45) {
-    printf("Temperature out of range!\n");
-    return 0;
-  } else if(soc < 20 || soc > 80) {
-    printf("State of Charge out of range!\n");
-    return 0;
-  } else if(chargeRate > 0.8) {
-    printf("Charge Rate out of range!\n");
-    return 0;
-  }
-  return 1;
+#include "checker.h"
+
+float Temperature;
+float StateOfCharge;
+float ChargeRate;
+
+float convertFarenheitToCelcius(float temp)
+{
+    return (temp - 32.0) * (5.0 / 9.0);
+}
+
+ParamLevel getInLimitParamLevel(float paramValue, const ParamAttributes * param)
+{
+    float tolerance = param->max * INLIMIT_TOLERANCE_PERCENTAGE;
+    ParamLevel level;
+    
+    if (paramValue <= (param->min + tolerance))
+    {
+        level = LOW_WARNING;
+    }
+    else if (paramValue >= (param->max - tolerance))
+    {
+        level = HIGH_WARNING;
+    }
+    else
+    {
+        level = NORMAL;
+    }
+    return level;
+}
+
+ParamLevel getOutLimitParamLevel(float paramValue, const ParamAttributes * param)
+{
+    ParamLevel level;
+    
+    if (paramValue < (param->min))
+    {
+        level = LOW_BREACH;
+    }
+    else
+    {
+        level = HIGH_BREACH;
+    }
+
+    return level;
+}
+
+void ReportWarningMessage(ParamLevel level, const ParamAttributes * param)
+{
+    char warnMessage[50];
+    sprintf(warnMessage, "'%s' is in '%s' range!\n", param->paramName, WarningMessage[level]);
+    printf("%s", warnMessage);
+    return;
 }
 
-int main() {
-  assert(batteryIsOk(25, 70, 0.7));
-  assert(!batteryIsOk(50, 85, 0));
+int IsWithinLimit(float paramValue, const ParamAttributes * param)
+{
+    if ((paramValue < param->min) || (paramValue > param->max))
+    {
+        char outMessage[50];
+        sprintf(outMessage, "%s = %0.4f, out of range!\n", param->paramName, paramValue);
+        printf("%s", outMessage);
+        ReportWarningMessage(getOutLimitParamLevel(paramValue, param), param);
+        return NOT_IN_LIMIT;
+    }
+    else
+    {
+        ReportWarningMessage(getOutLimitParamLevel(paramValue, param), param);
```

Comment:
Both the if and the else are calling `getOutLimitParamLevel`.
Write a test to expose the error before fixing it :)

Reply:
added test case and then corrected code too